lms/services/digest.py

```diff
@@ -0,0 +1,264 @@
+
+    @property
+    def display_name(self) -> Optional[str]:
+        """Return a display name for this user."""
+        for user in self.users:
+            if user.display_name:
+                return user.display_name
+
+        return None
+
+    @classmethod
+    def make(cls, db, h_userid):
+        """Return a UnifiedUser for the given h_userid."""
+        return cls(
+            db.scalars(
+                select(User).filter_by(h_userid=h_userid).order_by(User.updated.desc())
+            ).all()
+        )
+
+
+@dataclass
+class UnifiedCourse:
+    """All Course's for a given authority_provided_id, unified across all ApplicationInstance's."""
+
+    courses: List[Course] = field(default_factory=list)
+    annotations: List[dict] = field(default_factory=list)
+
+    def __post_init__(self):
+        for course in self.courses:
+            assert course.authority_provided_id == self.authority_provided_id
+
+    @property
+    def authority_provided_id(self) -> str:
+        """Return this course's authority_provided_id."""
+        return self.courses[0].authority_provided_id
+
+    @property
+    def title(self) -> str:
+        """Return a title for this course."""
+        return self.courses[0].lms_name
+
+    def learner_annotations(self, db) -> List[dict]:
+        """Return a list of all learner annotations in this course."""
+        return [
+            annotation
+            for annotation in self.annotations
+            if not self.is_instructor(
+                db, UnifiedUser.make(db, annotation["author"]["userid"])
+            )
+        ]
+
+    def is_instructor(self, db, user: UnifiedUser) -> bool:
+        """Return True if `user` is an instructor in this course."""
+        course_ids = [course.id for course in self.courses]
+
+        return bool(
+            db.scalar(
+                select(func.count())
+                .select_from(AssignmentMembership)
+                .join(Assignment)
+                .join(AssignmentGrouping)
+                .join(LTIRole)
+                .filter(AssignmentGrouping.grouping_id.in_(course_ids))
+                .filter(AssignmentMembership.user_id.in_(user.user_ids))
+                .filter(LTIRole.type == "instructor", LTIRole.scope == "course")
+            )
+        )
+
+    @staticmethod
+    def get_courses(db, authority_provided_id):
+        """Return a list of all Course's with the given authority_provided_id."""
+        return db.scalars(
+            select(Course)
+            .filter_by(authority_provided_id=authority_provided_id)
+            .order_by(Course.updated.desc())
+        ).all()
+
+    @classmethod
+    def make(cls, db, authority_provided_id):
+        """Return a UnifiedCourse for the given authority_provided_id."""
+        first_grouping = db.scalars(
+            select(Grouping)
+            .filter_by(authority_provided_id=authority_provided_id)
+            .order_by(Grouping.updated.desc())
+        ).first()
```

Comment:
I reckon this `first` deserves a comment on the assumptions it's making.

Reply:
Added a comment: https://github.com/hypothesis/lms/pull/5138/commits/19a33dded9e0b000589625ec4dd2a9845bb1f1fb